core/common/src/main/java/alluxio/exception/DirectoryNotInSyncException.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.exception;
+
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * Exception for situations where an operation fails because a directory is not in sync with the
+ * UFS. An example of this would be deleting a nonempty directory which has UFS contents not loaded
+ * into Alluxio.
+ */
+@ThreadSafe
+public class DirectoryNotInSyncException extends AlluxioException {
```

Comment:
@madanadit could we instead use `FailedPreconditionException` for this error case? It's a subclass of `IOException` so there would be very few changes from the original code